Utilize fingerprint as part of the resource

esanmiguelc opened the pull request:
This article points out the motivation for these changes:
https://www.stevesouders.com/blog/2008/08/23/revving-filenames-dont-use-querystring/

 - Adds an extra task argument that takes a vector of file extensions
   to be replaced by fingerprint
    - if none specified it defaults to `[".js" ".css"]`

esanmiguelc committed:
Utilize fingerprint as part of the resource

This article points out the motivation for these changes:
https://www.stevesouders.com/blog/2008/08/23/revving-filenames-dont-use-querystring/

 - Adds an extra task argument that takes a vector of file extensions
   to be replaced by fingerprint
    - if none specified it defaults to [".js" ".css"]

esanmiguelc commented:
@AdamFrey I would also consider making this a `2.0.0` release, this implementation moves the files around which might come unexpected to someone making an upgrade. I don't know if this would break some peoples' workflows or if this is a non-issue. I'm open to suggestions.

RadicalZephyr commented:
One thing you could do that might make it feel less necessary to bump to 2.0 would be to also introduce a fingerprinting-strategy option that defaults to query-string.  Then the old behavior would still be the default and people can upgrade and opt-in to the new behavior when they want to.

AdamFrey commented on src/afrey/boot_asset_fingerprint.clj:
```diff
@@ -27,8 +36,9 @@
-                          (map (juxt :path :hash))
-                          (core/output-files fileset))]
+                              (map (juxt :path :hash))
+                              (core/by-ext (or rename-extensions [".js" ".css"]) (core/output-files fileset)))
```
What's the reasoning for adding the rename-extensions list? Is it a performance tweak, because of the new `file-rename-hash` calculation? I worry that if someone adds the `${}` selector around, say an `img` resource, but they forget to add `jpeg` to the extension list, they will be confused when their resource isn't found.

esanmiguelc replied:
This implementation calls `core/mv` https://github.com/AdamFrey/boot-asset-fingerprint/pull/3/files/4431d63d72e46d6f39a50550be4a188bff394698#diff-62eeaa3ab1757be261299d76e96ebb0fR62 on the files in the hash, In order to avoid accidental fingerprinting of files such as a `.clj` the `rename-extensions` made sense.

RadicalZephyr replied:
It would add significant complexity, but it might be interesting to only rename files that actually had a fingerprint replacement done in some other file (i.e. that filename was found inside of a `${}`).

Then there would be no need for `rename-extensions` and fingerprinting any file would __Just Work__ ™️.

esanmiguelc replied:
I like that idea, 
I was thinking something like:
```markdown
1. Evaluating the files captured by `extensions`
2. Collect file paths
3. Filter files from the fileset and fingerprint
4. Replace template values with fingerprinted version
```
Does that sound like a reasonable approach? @RadicalZephyr @AdamFrey

AdamFrey replied:
Yes, that sounds like the right track to me. 👍

AdamFrey commented:
That's what I'm thinking. I can do some work to that effect this weekendOn Feb 16, 2017 12:39 PM, Geoff Shannon <notifications@github.com> wrote:@RadicalZephyr commented on this pull request.



In src/afrey/boot_asset_fingerprint.clj:
> @@ -27,8 +36,9 @@
                           (core/output-files)
                           (core/by-ext (or extensions [".html"])))
             file-hashes (into {}
-                          (map (juxt :path :hash))
-                          (core/output-files fileset))]
+                              (map (juxt :path :hash))
+                              (core/by-ext (or rename-extensions [".js" ".css"]) (core/output-files fileset)))

It would add significant complexity, but it might be interesting to only rename files that actually had a fingerprint replacement done in some other file (i.e. that filename was found inside of a ${}).
Then there would be no need for rename-extensions and fingerprinting any file would Just Work ™️.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

lsnape commented:
I'd love to use this task for my production builds but can't until this feature is merged. Anything I can do to help hurry it along?

AdamFrey commented:
I haven't heard anything from @esanmiguelc in a while, but I might take a crack at getting this done this weekend

AdamFrey committed:
Move test code to test/src

AdamFrey committed:
Add function for pulling all desired assets from sources

AdamFrey committed:
Pull out and name default extension lists

AdamFrey committed:
Fix resource-paths typo

AdamFrey committed:
Determine files to fingerprint programatically from source files

This way we can avoid the rename-extensions option. See the discussion here:
https://github.com/AdamFrey/boot-asset-fingerprint/pull/3/files/4431d63d72e46d6f39a50550be4a188bff394698#r101436618

AdamFrey committed:
Fix indentation

AdamFrey commented:
deployed as `1.3.0`. Thanks @esanmiguelc
